app/services/foreman_templates/template_exporter.rb

```diff
@@ -107,5 +107,41 @@ def export_method
           raise "Unknown metadata export mode #{@metadata_export_mode}"
       end
     end
+
+    private
+
+    def find_templates
+      @organizations.empty? && @locations.empty? ? Template.all : find_taxed_templates
+    end
+
+    def find_taxed_templates
+      Template.subclasses.reduce([]) do |memo, subclass|
+        memo.concat templates_in_tax(subclass)
+      end
+    end
+
+    def templates_in_tax(template_class)
+      location_ids = taxes_ids('location')
+      organization_ids = taxes_ids('organization')
```

Comment:
you can keep ids in a single array, their ids will always be unique, e.g. `taxonomy_ids = location_ids + organization_ids`

then below it does not really matter what model do you use, you could just do
```
templates_query(taxable_ids)
```
where templates_query would be using `Taxonomy` parent directly, e.g.
```
Taxonomy.join(:taxable_taxonomies).where(:taxable_taxonomies => { :taxonomy_id => tax_ids })
```

Rails will correctly load the type when it builds instances.